File: app/te2_paths.py
from __future__ import annotations

import os
import stat
import tempfile
from collections.abc import Callable, Mapping, MutableMapping
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class Te2Paths:
    cache_home: Path
    data_home: Path
    config_home: Path
    runtime_home: Path

    def export(self, environ: MutableMapping[str, str]) -> None:
        environ["TE2_CACHE_HOME"] = str(self.cache_home)
        environ["TE2_DATA_HOME"] = str(self.data_home)
        environ["TE2_CONFIG_HOME"] = str(self.config_home)
        environ["TE2_RUNTIME_HOME"] = str(self.runtime_home)


def te2_cache_home(
    environ: Mapping[str, str] | None = None,
    *,
    home: Path | None = None,
) -> Path:
    return _persistent_home(
        environ if environ is not None else os.environ,
        explicit_key="TE2_CACHE_HOME",
        xdg_key="XDG_CACHE_HOME",
        fallback=lambda: _user_home(environ, home) / ".cache",
    )


def te2_data_home(
    environ: Mapping[str, str] | None = None,
    *,
    home: Path | None = None,
) -> Path:
    return _persistent_home(
        environ if environ is not None else os.environ,
        explicit_key="TE2_DATA_HOME",
        xdg_key="XDG_DATA_HOME",
        fallback=lambda: _user_home(environ, home) / ".local" / "share",
    )


def te2_config_home(
    environ: Mapping[str, str] | None = None,
    *,
    home: Path | None = None,
) -> Path:
    return _persistent_home(
        environ if environ is not None else os.environ,
        explicit_key="TE2_CONFIG_HOME",
        xdg_key="XDG_CONFIG_HOME",
        fallback=lambda: _user_home(environ, home) / ".config",
    )


def te2_runtime_home(
    environ: Mapping[str, str] | None = None,
    *,
    uid: int | None = None,
    platform_temp: Path | None = None,
) -> Path:
    source = environ if environ is not None else os.environ
    explicit = _nonempty(source.get("TE2_RUNTIME_HOME"))
    if explicit:
        return _absolute_path(explicit, "TE2_RUNTIME_HOME")

    xdg_runtime = _nonempty(source.get("XDG_RUNTIME_DIR"))
    if xdg_runtime:
        return _absolute_path(xdg_runtime, "XDG_RUNTIME_DIR") / "te2"

    temporary = _nonempty(source.get("TMPDIR"))
    if temporary:
        temporary_root = _absolute_path(temporary, "TMPDIR")
    elif _is_termux(source) and _nonempty(source.get("PREFIX")):
        temporary_root = _absolute_path(str(source["PREFIX"]), "PREFIX") / "tmp"
    else:
        temporary_root = Path(platform_temp or tempfile.gettempdir())
        if not temporary_root.is_absolute():
            raise ValueError(f"platform temporary directory must be absolute: {temporary_root}")

    resolved_uid = uid if uid is not None else _current_uid()
    return temporary_root / f"te2-{resolved_uid}"


def resolve_te2_paths(
    environ: Mapping[str, str] | None = None,
    *,
    home: Path | None = None,
    uid: int | None = None,
    platform_temp: Path | None = None,
) -> Te2Paths:
    source = environ if environ is not None else os.environ
    return Te2Paths(
        cache_home=te2_cache_home(source, home=home),
        data_home=te2_data_home(source, home=home),
        config_home=te2_config_home(source, home=home),
        runtime_home=te2_runtime_home(source, uid=uid, platform_temp=platform_temp),
    )


def ensure_runtime_home(path: Path, *, uid: int | None = None) -> Path:
    if not path.is_absolute():
        raise ValueError(f"TE2 runtime root must be absolute: {path}")
    if path.is_symlink():
        raise RuntimeError(f"TE2 runtime root must not be a symbolic link: {path}")

    path.mkdir(mode=0o700, parents=True, exist_ok=True)
    if path.is_symlink():
        raise RuntimeError(f"TE2 runtime root must not be a symbolic link: {path}")

    metadata = path.stat()
    if not stat.S_ISDIR(metadata.st_mode):
        raise RuntimeError(f"TE2 runtime root is not a directory: {path}")

    expected_uid = uid if uid is not None else _current_uid()
    if hasattr(metadata, "st_uid") and metadata.st_uid != expected_uid:
        raise RuntimeError(
            f"TE2 runtime root is owned by uid {metadata.st_uid}, expected {expected_uid}: {path}"
        )

    if stat.S_IMODE(metadata.st_mode) != 0o700:
        path.chmod(0o700)
    return path


def _persistent_home(
    environ: Mapping[str, str],
    *,
    explicit_key: str,
    xdg_key: str,
    fallback: Callable[[], Path],
) -> Path:
    explicit = _nonempty(environ.get(explicit_key))
    if explicit:
        return _absolute_path(explicit, explicit_key)

    xdg_base = _nonempty(environ.get(xdg_key))
    if xdg_base:
        return _absolute_path(xdg_base, xdg_key) / "te2"

    fallback_root = fallback()
    if not fallback_root.is_absolute():
        raise ValueError(f"TE2 fallback root must be absolute: {fallback_root}")
    return fallback_root / "te2"


def _user_home(environ: Mapping[str, str] | None, override: Path | None) -> Path:
    if override is not None:
        home = override
    else:
        source = environ if environ is not None else os.environ
        raw_home = _nonempty(source.get("HOME"))
        home = _absolute_path(raw_home, "HOME") if raw_home else Path.home()
    if not home.is_absolute():
        raise ValueError(f"HOME must be absolute: {home}")
    return home


def _absolute_path(raw: str, key: str) -> Path:
    path = Path(raw)
    if not path.is_absolute():
        raise ValueError(f"{key} must be an absolute path: {raw!r}")
    return path


def _nonempty(value: str | None) -> str:
    return str(value or "").strip()


def _is_termux(environ: Mapping[str, str]) -> bool:
    prefix = _nonempty(environ.get("PREFIX"))
    return bool(
        _nonempty(environ.get("ANDROID_ROOT"))
        or _nonempty(environ.get("ANDROID_DATA"))
        or "/com.termux/" in prefix
    )


def _current_uid() -> int:
    getuid = getattr(os, "getuid", None)
    if not callable(getuid):
        return 0
    uid = getuid()
    if not isinstance(uid, int):
        raise RuntimeError(f"os.getuid() returned a non-integer value: {uid!r}")
    return uid

File: desktop_client/electron/src/main/te2-paths.ts
import { chmod, lstat, mkdir } from "node:fs/promises";
import { homedir, tmpdir } from "node:os";
import { isAbsolute, join } from "node:path";

export type Te2Paths = {
  cacheHome: string;
  dataHome: string;
  configHome: string;
  runtimeHome: string;
};

function nonempty(value: string | undefined): string {
  return String(value || "").trim();
}

function absolutePath(value: string, key: string): string {
  if (!isAbsolute(value)) throw new Error(`${key} must be an absolute path: ${JSON.stringify(value)}`);
  return value;
}

function userHome(environment: NodeJS.ProcessEnv, override?: string): string {
  return absolutePath(override || nonempty(environment.HOME) || homedir(), "HOME");
}

function persistentHome(
  environment: NodeJS.ProcessEnv,
  explicitKey: "TE2_CACHE_HOME" | "TE2_DATA_HOME" | "TE2_CONFIG_HOME",
  xdgKey: "XDG_CACHE_HOME" | "XDG_DATA_HOME" | "XDG_CONFIG_HOME",
  fallbackBase: () => string,
): string {
  const explicit = nonempty(environment[explicitKey]);
  if (explicit) return absolutePath(explicit, explicitKey);
  const xdg = nonempty(environment[xdgKey]);
  if (xdg) return join(absolutePath(xdg, xdgKey), "te2");
  return join(fallbackBase(), "te2");
}

export function te2CacheHome(environment = process.env, home?: string): string {
  return persistentHome(
    environment,
    "TE2_CACHE_HOME",
    "XDG_CACHE_HOME",
    () => join(userHome(environment, home), ".cache"),
  );
}

export function te2DataHome(environment = process.env, home?: string): string {
  return persistentHome(
    environment,
    "TE2_DATA_HOME",
    "XDG_DATA_HOME",
    () => join(userHome(environment, home), ".local", "share"),
  );
}

export function te2ConfigHome(environment = process.env, home?: string): string {
  return persistentHome(
    environment,
    "TE2_CONFIG_HOME",
    "XDG_CONFIG_HOME",
    () => join(userHome(environment, home), ".config"),
  );
}

export function te2RuntimeHome(
  environment = process.env,
  platformTemp = tmpdir(),
  uid = process.getuid?.() ?? 0,
): string {
  const explicit = nonempty(environment.TE2_RUNTIME_HOME);
  if (explicit) return absolutePath(explicit, "TE2_RUNTIME_HOME");
  const xdg = nonempty(environment.XDG_RUNTIME_DIR);
  if (xdg) return join(absolutePath(xdg, "XDG_RUNTIME_DIR"), "te2");

  const environmentTemp = nonempty(environment.TMPDIR);
  let temporaryRoot: string;
  if (environmentTemp) {
    temporaryRoot = absolutePath(environmentTemp, "TMPDIR");
  } else if (
    nonempty(environment.ANDROID_ROOT) ||
    nonempty(environment.ANDROID_DATA) ||
    nonempty(environment.PREFIX).includes("/com.termux/")
  ) {
    const prefix = nonempty(environment.PREFIX);
    temporaryRoot = prefix
      ? join(absolutePath(prefix, "PREFIX"), "tmp")
      : absolutePath(platformTemp, "platform temporary directory");
  } else {
    temporaryRoot = absolutePath(platformTemp, "platform temporary directory");
  }
  return join(temporaryRoot, `te2-${uid}`);
}

export function resolveTe2Paths(
  environment = process.env,
  options: { home?: string; platformTemp?: string; uid?: number } = {},
): Te2Paths {
  return {
    cacheHome: te2CacheHome(environment, options.home),
    dataHome: te2DataHome(environment, options.home),
    configHome: te2ConfigHome(environment, options.home),
    runtimeHome: te2RuntimeHome(environment, options.platformTemp, options.uid),
  };
}

export async function ensureRuntimeHome(path: string, uid = process.getuid?.() ?? 0): Promise<string> {
  absolutePath(path, "TE2 runtime root");
  await mkdir(path, { recursive: true, mode: 0o700 });
  let metadata = await lstat(path);
  if (metadata.isSymbolicLink()) throw new Error(`TE2 runtime root must not be a symbolic link: ${path}`);
  if (!metadata.isDirectory()) throw new Error(`TE2 runtime root is not a directory: ${path}`);
  if (typeof metadata.uid === "number" && metadata.uid !== uid) {
    throw new Error(`TE2 runtime root is owned by uid ${metadata.uid}, expected ${uid}: ${path}`);
  }
  if ((metadata.mode & 0o777) !== 0o700) {
    await chmod(path, 0o700);
    metadata = await lstat(path);
    if ((metadata.mode & 0o777) !== 0o700) throw new Error(`TE2 runtime root permissions are not 0700: ${path}`);
  }
  return path;
}
